crates/db_views/src/post_view.rs

```diff
@@ -417,12 +429,43 @@ impl PostView {
   }
 }
 
-#[derive(Default)]
+impl PaginationCursor {
+  // get cursor for page after the given posts
+  pub fn after(posts: &[PostView]) -> Option<PaginationCursor> {
+    posts
+      .last()
+      .map(|p| PaginationCursor(format!("P{:x}", p.post.id.0)))
+  }
+  pub async fn read(&self, pool: &mut DbPool<'_>) -> Result<PaginationCursorData, Error> {
```

Comment:
The post needs to be read because for filtering I'm not actually using the current post ID directly, but rather a subset of properties of that post. The reason this is needed is that we need a multi-part WHERE filter that exactly matches each index we have for each separate sorting method in order for postgres to be able to optimize it.

For example, for the hot sort, it doesn't do `WHERE id > pagination_token_post.id` but rather it does `WHERE hot_rank <= pagination_token_post.hot_rank AND featured_local <= pagination_token_post.featured_local AND published <= pagination_token_post.published OFFSET 1`

The exact set of parameters we actually need in order to paginate differs for each sort type, but the easiest way (for now) is to just use a post id as a proxy for those values.

Reply:
All the cursor-based-paging examples I can find, use simple and predictable id or date columns. Since our most important sorts don't use those, and use multiple columns, this is going to get really complex *really fast*, especially as we add more sort and filter columns. We have to keep this simple and maintainable.

My main questions are:

- What does a compound-column cursor actually look like, and how does it work? I can't find any examples online.
- Are cursors *only* based on the sort column? Or are they also based on the filters / joins?
- If cursors are *only* based on sorts, wouldn't it maybe be easier to alter or add columns to these aggregates tables, (and make scheduled tasks to update these columns accordingly). What you suggested in #2877 , which was to add columns like `community_hot_rank_position`, seems like it would work really well. 
  - If it uses filters, then this might be impossible: post_view has A LOT of filters. The number of`order` columns could explode, but even that *still* might make more sense, and be simpler, and more predictable.

Reply:
> this is going to get really complex really fast

Well, it's not really going to get more complex when using PostId as a proxy for the compound cursor, since the given post defines all properties we could want to sort by (even in the future). No code changes are required for adding more sorts, for example your `scaled` sort should be addable without any changes specific to this pagination.

1. A compound cursor always just defines the start of the page (e.g. `featured_local <=, hot_rank <=, published <=`). The sort order (ORDER BY featured_local desc, hot_rank desc, published_desc) is still determined by the sort type parameter. But if you use a cursor from one sort for a different sort, you'll get unexpected results.
2. They are only based on the sort column. If your query filters out more stuff, then the second-page cursor will be different, but just because it's further down the list of posts.
3. Sure, that would be possible, but you need one of those columns for every possible sort, and you need those maintenance tasks. The code would also be more complex and longer (imo). The neat thing about this solution (if it scales as expected), is that no further code changes are required and it will work with any new sort we might add. It's just a bit harder to understand.

Reply:
Since the cursor is not actually *just* the post id, but also the extra fields necessary for sorting, wouldn't it be easier to include those in the `PaginationCursor` ? You can serialize / deserialize them however you like, but that would avoid the extra calls.

Reply:
(similar answer to the below discussion)

The issue with that is that either it would mean the pagination cursor would be huge (the whole post aggregates object), or there would need to be many special cases for every kind of sort (since every sort needs a different subset of information for the pagination column). it would also require some compact url-safe encoding.

Trying the simplest solution (postaggregates -> json+base64 encoding) results in pagination tokens that are almost 1000 characters long which is a lot for a URL parameter.

I don't think right now the extra post fetch is much of an issue considering all the other stuff this request does is 100+ times as expensive. But that's also the reason i made the token into a visibly opaque string so we can change this freely later (no breaking change required)